### ISSUE TITLE
OpenMM ForceField handling for SMIRNOFF impropers

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#2509]
I'm finishing up [this `openmm-forcefields` PR](https://github.com/choderalab/openmm-forcefields/pull/79) to add both GAFF and SMIRNOFF residue template generators for the OpenMM `ForceField` facility, and having a bit of trouble getting `ForceField` to replicate the [SMIRNOFF improper behavior](https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html#impropertorsions) in which all three impropers around a trefoil are added. That is, if we have `i-j(k)-l`, where `j` is the central atom, all three torsions (`j-i-k-l`, `j-k-l-i`, `j-l-i-k`) are added. 

The SMIRNOFF residue template generator creates a unique atom type for each atom of the small molecule and then generates improper terms for each of these impropers, but only one improper seems to be added to the system. I'm trying to make sense of the [`_matchImproper`](https://github.com/openmm/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L1744) code to see if one of the other modes besides `default` may be more appropriate, but it might be the case that I need to add a `smirnoff` mode here. 

Any ideas if there is a trick I could use to exploit one of the existing modes, or any pointers for adding a new mode to ensure all three are added?

**Comment by peastman** [comment]
I think this will require adding a new mode.  `_matchImproper()` gets called once for each cluster of three atoms around a central atom.  It uses `itertools.permutations()` to loop over all possible permutations of the three atoms and returns at most a single match.  We'll need to allow it to return multiple matches.

**Comment by peastman** [comment]
You also might consider using a different functional form.  "Improper torsions" were a hack to begin with.  People needed to make groups of four atoms lie in a plane.  They'd already written code for torsions, and realized they could repurpose it.  But it isn't symmetric with respect to the atom order, which creates problems.  Adding multiple impropers is a hack on top of that hack to try to symmetrize it.  But only adding three isn't enough.  If you really want it to be symmetric, you need to add six.  For example, j-i-k-l and j-i-l-k have different angles.

A better alternative would be to use a functional form that's symmetric from the beginning.  For example, you could use the distance of the central atom from the plane of the other three.  This has several advantages:

- It's totally symmetric, so the atom order doesn't matter.
- It's faster to calculate.
- It's much better behaved numerically.  It's continuous and well behaved everywhere, in contrast to torsions that have singularities intrinsic to the function (the angle is undefined when three atoms are in a line) and others that come from the algorithm used to calculate them (the singularities in acos).

**Comment by jchodera** [comment]
> I think this will require adding a new mode.

@peastman I was able to address this in https://github.com/openmm/openmm/pull/2511 via a minimal change that does not require returning multiple matches.

> You also might consider using a different functional form. "Improper torsions" were a hack to begin with. People needed to make groups of four atoms lie in a plane. They'd already written code for torsions, and realized they could repurpose it. But it isn't symmetric with respect to the atom order, which creates problems. Adding multiple impropers is a hack on top of that hack to try to symmetrize it. 

This is exactly the problem we intended to solve by having the [`ImproperTorsions` SMIRNOFF spec](https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html#impropertorsions) dictate that three torsions be added in a trefoil.

> But only adding three isn't enough. If you really want it to be symmetric, you need to add six. For example, j-i-k-l and j-i-l-k have different angles.

In fact, we originally added all six, but realized we quickly ran into sign issues. For pyramidal impropers where `j` is displaced above the `i-k-l` plane, the sign would be + for all `j-i-k-l`, `j-k-l-i`, and `j-l-i-k` torsions, but - for `j-i-l-k`, `j-l-k-i`, and `j-k-i-l` torsions. To solve this problem, we decide to keep only three torsions with consistent signs and apply them in a consistent pattern for simplicity of porting over old AMBER/GAFF torsions into a similar framework in a manner that can be supported in all major codes.

> A better alternative would be to use a functional form that's symmetric from the beginning. For example, you could use the distance of the central atom from the plane of the other three. 

100% agree! The [Open Force Field Initiative](http://openforcefield.org) is on the hook to deliver functional forms that are compatible in all major packages (OpenMM, AMBER, CHARMM, NAMD, gromacs) for the next year or so, but after that, we'll be better positioned to drive functional form innovation in other codes (provided we can also push for their broad adoption through standards). We're wide-open to suggestions like this that we can [add support for in new versions of the SMIRNOFF spec](https://open-forcefield-toolkit.readthedocs.io/en/0.6.0/smirnoff.html) and easily assess/use in OpenMM, but we'll need a way to drive adoption in other codes as well.
